### PR TITLE
[Internal Manager] Check allowance before asking to approve

### PR DESCRIPTION
**Pull request by luizakp**
https://linear.app/bleu-llc/issue/BAL-493/check-allowance-before-asking-to-approve
https://www.loom.com/share/f93700ad3e154948ab9698add5102cb7?sid=f6590b42-25d4-49fe-9010-1b9e209ed4e7

Todo:
- [ ] fetch all allowances on the first render and use it instead of reading the contract on every render
- [ ] clean the code/make more readable

### CODE REVIEW

File: packages/next-app/src/hooks/useTransaction.ts [hunk 1]
```diff
@@ -274,36 +275,75 @@ export function useInternalBalancesTransaction({
...
+      args: [userAddress, vaultAddress[5]],
     });
-    const data = await writeContract(config);
-    const { hash, wait } = data;
-    handleTransactionStatus({ hash });
-    const receipt = await wait();
-    if (receipt.status) {
+    const amountToApprove = parseFixed(tokenAmount, tokenDecimals);
+    if (allowance.gte(amountToApprove)) {
+      setHasAllowance(true); // User has enough allowance
```
Comment (mendesfabio) [inline packages/next-app/src/hooks/useTransaction.ts:302]: if you re-name `hasAllowance` to be more explicit maybe the comment isn't even needed, some examples: `isApprovalRequired` `hasEnoughAllowance` etc

**Review by mendesfabio — COMMENTED** [review]
good progress, don't have much to comment beside the suggestion to only fetch allowance on load and maybe `multicall` over the token list.

File: packages/next-app/src/hooks/useTransaction.ts [hunk 2]
```diff
@@ -343,6 +383,24 @@ export function useInternalBalancesTransaction({
...
+      setNotification(
+        NOTIFICATION_MAP_INTERNAL_BALANCES[TransactionStatus.WRITE_ERROR]
+      );
+      setTransactionStatus(TransactionStatus.CONFIRMING);
+    } else {
+      setNotification(
+        NOTIFICATION_MAP_INTERNAL_BALANCES[TransactionStatus.WRITE_ERROR]
+      );
+      setTransactionStatus(TransactionStatus.AUTHORIZING);
+    }
```
Comment (ribeirojose) [inline packages/next-app/src/hooks/useTransaction.ts:401]: Can you elaborate on why we need to set the transaction status after erroring out?

File: packages/next-app/src/hooks/useTransaction.ts [hunk 3]
```diff
@@ -274,36 +275,75 @@ export function useInternalBalancesTransaction({
...
+        NOTIFICATION_MAP_INTERNAL_BALANCES[TransactionStatus.WAITING_APPROVAL]
       );
+      const config = await prepareWriteContract({
+        address: submitData[0].tokenAddress as Address,
+        abi: erc20ABI,
+        functionName: "approve",
+        args: [
+          vaultAddress[5],
+          parseFixed(
+            submitData[0]?.tokenAmount ? submitData[0].tokenAmount : "0",
```
Comment (ribeirojose) [inline packages/next-app/src/hooks/useTransaction.ts:326]: wouldn't `submitData[0]?.tokenAmount || "0",` work?

File: packages/next-app/src/hooks/useTransaction.ts (hunk 2 above)
Comment (luizakp) [reply packages/next-app/src/hooks/useTransaction.ts:401]: So the user can reject on metamask and still be able to try again https://www.loom.com/share/32d8aad314204ed3922d10062f4ba639?sid=46aefd58-a45b-4944-af3d-41e26e516747

**Pull request merged**